## Gather

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [3]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [4]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [5]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [9]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [10]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [11]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [12]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [13]:
patients.sample(5)

patient_id assigned_sex given_name   surname             address  \
9            10       female     Sophie   Cabrera   3303 Anmoore Road   
288         289         male     Fareeq      Kalb  1043 Ventura Drive   
495         496         male     Hajime   Tsukada   4111 Thunder Road   
165         166         male     Zlatko  Rukavina       592 Rafe Lane   
282         283       female      Sandy    Taylor  2476 Fulton Street   

           city       state  zip_code        country  \
9      New York    New York   10011.0  United States   
288  Santa Cruz  California   95060.0  United States   
495   San Mateo          CA   94403.0  United States   
165  Yazoo City          MS   39194.0  United States   
282    Rainelle          WV   25962.0  United States   

                                           contact   birthdate  weight  \
9    SophieCabreraIbarra@teleworm.us1 718 795 9124   12/3/1930   194.7   
288       831-427-4114FareeqJawharKalb@armyspy.com    5/1/1985   194.5   
495           650-570-4896HajimeTsukada@dayrep.com    9/5/1972   168.1   
165       ZlatkoRukavina@cuvox.de+1 (662) 716-9586  12/12/1976   198.7   
282           304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1   

     height   bmi  
9        64  33.4  
288      70  27.9  
495      66  27.1  
165      70  28.5  
282      64  35.4

In [14]:
patients['surname'].value_counts()

Doe           6
Taylor        3
Jakobsen      3
Ogochukwu     2
Souza         2
             ..
Wubbels       1
Wiśniewska    1
Fejes         1
Baum          1
Wilkinson     1
Name: surname, Length: 466, dtype: int64

In [15]:
patients['address'].value_counts()

123 Main Street         6
2476 Fulton Street      2
2778 North Avenue       2
648 Old Dear Lane       2
3958 Liberty Avenue     1
                       ..
1386 Farland Street     1
2272 Williams Avenue    1
204 Rockwell Lane       1
4380 Riverside Drive    1
1109 Beechwood Drive    1
Name: address, Length: 483, dtype: int64

In [26]:
patients[patients['surname'].duplicated()]

patient_id assigned_sex  given_name      surname  \
26           27       female        Ella         Lund   
29           30         male        Jake     Jakobsen   
122         123         male         Bội           Tạ   
139         140       female     Novalie         Berg   
194         195         male        Urso       Aranda   
195         196       female      Lamara     Dratchev   
203         204       female         Mùi        Lương   
205         206         male       Breno      Correia   
207         208       female   Beatrycze      Woźniak   
229         230         male        John          Doe   
235         236       female     Justyna    Kowalczyk   
237         238         male        John          Doe   
244         245         male        John          Doe   
251         252         male        John          Doe   
277         278         male        John          Doe   
282         283       female       Sandy       Taylor   
295         296         male      Albino    Schiavone   
296         297       female         Chỉ          Lâm   
304         305       female       Signý  Grímsdóttir   
308         309         male        Daud    Batukayev   
317         318       female       Nancy       Parker   
322         323         male        Mười         Liễu   
336         337       female      Sharon      Collins   
350         351         male        Diệt          Bùi   
364         365       female  Chukwumoge    Ogochukwu   
407         408       female       Tegan      Johnson   
418         419         male      Mahmud      Kadyrov   
426         427         male     Rogelio       Taylor   
431         432       female     Manuela      Cindrić   
432         433       female       Karen     Jakobsen   
441         442         male        Luis        Silva   
449         450         male     Bjørnar       Nilsen   
461         462         male      Cannan      Cabrera   
469         470         male      Harley       Tucker   
482         483         male       Diogo        Souza   
496         497         male   Alexander       Hueber   
502         503         male         Pat      Gersten   

                      address              city       state  zip_code  \
26         1207 Garfield Road            Peoria          IL   61602.0   
29          648 Old Dear Lane       Port Jervis    New York   12771.0   
122        2389 Rubaiyat Road      Grand Rapids          MI   49503.0   
139          1275 Goldie Lane        Cincinnati          OH   45202.0   
194       1330 Lincoln Street  Hopewell  Mercer          NJ    8525.0   
195    3731 Swick Hill Street       New Orleans          LA   70113.0   
203        1778 Rodney Street         Harvester          MO   63301.0   
205          3539 Bottom Lane         Tonawanda    New York   14150.0   
207          182 Cross Street           Saginaw          MI   48607.0   
229           123 Main Street          New York          NY   12345.0   
235           3115 May Street      Pippa Passes          KY   41844.0   
237           123 Main Street          New York          NY   12345.0   
244           123 Main Street          New York          NY   12345.0   
251           123 Main Street          New York          NY   12345.0   
277           123 Main Street          New York          NY   12345.0   
282        2476 Fulton Street          Rainelle          WV   25962.0   
295    3113 Timber Ridge Road        Sacramento  California   95814.0   
296                       NaN               NaN         NaN       NaN   
304           3268 Karen Lane        Louisville          KY   40202.0   
308       153 Fieldcrest Road        Huntington    New York   11743.0   
317          4605 Hall Street         Las Vegas          NV   89110.0   
322         4346 Sussex Court           Chilton          TX   76632.0   
336   1240 Summit Park Avenue           Pontiac          MI   48342.0   
350    3788 Worthington Drive             Rylie          TX   75235.0   
364          

In [17]:
patients['weight'].sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [18]:
sum(treatments['auralin'].isnull())

0

In [19]:
sum(treatments['novodra'].isnull())

0

In [28]:
# Find duplicate column names
all_columns = pd.Series(list(patients) 
                        + list(treatments)
                        + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

### Quality
#### `patients` table
- zip code is a float not a string
- zip code has four digits sometimes
- Tim Neudorf height is 27 inches instead of 72 inches
- full state names sometimes, abbreviations other times
- mispelled Dsvid Gustafsson
- missing demographic information (address - contact columns)
- erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- multiple phone number formats
- default 'John Doe' data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

#### `treatments` table
- missing HbA1c changes
- the letter u in starting and ending doses for Auralin and Novodra
- missing records (280 instead of 350)
- erroneous datatype (auralin and novodra columns) 
- innacurate HbA1c changes (4s mistaken as 9s)
- nulls represented as dashes (-) in auralin and novodra columns

#### `adverse_reactions` table
- lowercase given names and surnames

### Tidiness
#### `patients` table
- contact column should be split into phone number and email

#### `treatments` table
- three variables in two columns (treatment: auralin and novodra, start dose, and end dose)